bluebutton-data-pipeline-rif-extract/src/main/java/gov/hhs/cms/bluebutton/datapipeline/rif/extract/s3/DataSetMonitorWorker.java

```diff
@@ -0,0 +1,283 @@
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3Client;
+import com.amazonaws.services.s3.model.DeleteObjectsRequest;
+import com.amazonaws.services.s3.model.DeleteObjectsRequest.KeyVersion;
+import com.amazonaws.services.s3.model.GetObjectRequest;
+import com.amazonaws.services.s3.model.ListObjectsRequest;
+import com.amazonaws.services.s3.model.ObjectListing;
+import com.amazonaws.services.s3.model.S3Object;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+
+import gov.hhs.cms.bluebutton.datapipeline.rif.extract.s3.DataSetManifest.DataSetManifestEntry;
+import gov.hhs.cms.bluebutton.datapipeline.rif.model.RifFile;
+import gov.hhs.cms.bluebutton.datapipeline.rif.model.RifFilesEvent;
+
+/**
+ * <p>
+ * Acts as a worker {@link Runnable} for {@link DataSetMonitor}. It is expected that
+ * this will be run on a repeating basis, via a {@link ScheduledExecutorService}
+ * .
+ * </p>
+ * <p>
+ * When executed via {@link #run()}, the {@link DataSetMonitorWorker} will scan the
+ * specified Amazon S3 bucket. It will look for <code>manifest.xml</code>
+ * objects/files and select the oldest one available. If such a manifest is
+ * found, it will then wait for all of the objects in the data set represented
+ * by it to become available. Once they're all available, it will kick off the
+ * processing of the data set, and block until that processing has completed.
+ * </p>
+ */
+final class DataSetMonitorWorker implements Runnable {
+	private static final Pattern REGEX_MANIFEST = Pattern.compile("^(.*)\\/manifest\\.xml$");
+
+	private final String bucketName;
+	private final DataSetProcessor dataSetProcessor;
+	private final AmazonS3 s3Client = new AmazonS3Client(new DefaultAWSCredentialsProviderChain());
+
+	/**
+	 * Constructs a new {@link DataSetMonitorWorker} instance.
+	 * 
+	 * @param bucketName
+	 *            the name of the AWS S3 bucket to monitor
+	 * @param dataSetProcessor
+	 *            the {@link DataSetProcessor} responsible for processing each
+	 *            of the {@link RifFilesEvent}s that are generated by this
+	 *            {@link DataSetMonitorWorker}
+	 */
+	public DataSetMonitorWorker(String bucketName, DataSetProcessor dataSetProcessor) {
+		this.bucketName = bucketName;
+		this.dataSetProcessor = dataSetProcessor;
+	}
+
+	/**
+	 * @see java.lang.Runnable#run()
+	 */
+	@Override
+	public void run() {
+		/*
+		 * Request a list of all objects in the configured bucket and directory.
+		 * (In the results, we'll be looking for the oldest manifest file, if
+		 * any.)
+		 */
+		ListObjectsRequest bucketListRequest = new ListObjectsRequest();
+		bucketListRequest.setBucketName(bucketName);
+
+		/*
+		 * S3 will return results in separate pages. Loop through all of the
+		 * pages, looking for the oldest manifest file that is available.
+		 */
+		String manifestToProcessKey = null;
+		ObjectListing objectListing = s3Client.listObjects(bucketListRequest);
```

Comment:
> I largely agree that we could. I just suspect that, over time, our buckets may start to accrue some manually-modified "cruft", e.g. "BadDataIShouldDeleteButHavent-2007-04-06T00:00Z/manifest.xml". 

Anything that gets manually modified like this should probably be moved to a separate bucket. 

> I think I disagree with you on this one: we should assume that the manifests and RIF files might finish uploading in any order, particularly if they all start uploading at the same time. And yeah, that might not be a problem for very long, but really really infrequent race condition errors are even worse to debug than more frequent ones.

I meant very long in general, not very long in relation to the other files. Like if a folder is sitting there for hours with no manifest file, that simply should not be possible. The manifest should just about always be the smallest file, and files uploaded to S3 can be uploaded in parallel, so it should not wait for a larger file to complete. 

There is more danger, I think, in a scenario where let's say you have 2 timestamped buckets out there, and bucket #1 has the earlier timestamp. Its manifest file fails to transfer for some reason, but bucket #2 has its manifest file and once all its files there, it begins to process. Bucket #1 then gets completely ignored.